sled-agent/src/bootstrap/agent/rack_ops.rs

```diff
@@ -0,0 +1,300 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Internal API for rack-level bootstrap agent operations.
+
+use super::Agent;
+use crate::bootstrap::http_entrypoints::RackOperationStatus;
+use crate::bootstrap::params::RackInitializeRequest;
+use crate::bootstrap::rss_handle::RssHandle;
+use crate::config::SidecarRevision;
+use crate::rack_setup::service::SetupServiceError;
+use std::mem;
+use std::sync::Arc;
+use std::sync::Mutex;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::error::TryRecvError;
+use uuid::Uuid;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RackInitId(pub Uuid);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RackResetId(pub Uuid);
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum ConcurrentRssAccess {
+    #[error("RSS is still initializating and cannot run concurrently")]
+    StillInitializing,
+    #[error("RSS failed to initialize: {message}")]
+    InitializationFailed { message: String },
+    #[error("RSS panicked while initializing")]
+    InitializationPanicked,
+    #[error("RSS is already initialized")]
+    AlreadyInitialized,
+    #[error("RSS is still resetting and cannot run concurrently")]
+    StillResetting,
+    #[error("RSS failed to reset: {message}")]
+    ResetFailed { message: String },
+    #[error("RSS panicked while resetting")]
+    ResetPanicked,
+    #[error("RSS is already reset")]
+    AlreadyReset,
+}
+
+pub(super) struct RssAccess {
+    // Note: The `Mutex` here is a std mutex, not a tokio mutex, and thus not
+    // subject to async cancellation issues (and also cannot be held across an
+    // `.await` point). We only keep it held long enough to perform quick
+    // operations: fill it in or read its current value.
+    status: Arc<Mutex<RssStatus>>,
+}
+
+impl RssAccess {
+    pub(super) fn new(initialized: bool) -> Self {
+        let status = if initialized {
+            RssStatus::InitializedOnStartup
+        } else {
+            RssStatus::ResetOnStartup
```

Comment:
Just to make sure I understand the semantics here:

- If we can't read necessary metadata from the M.2, we pass `initialized = false`
- This sets the state to `ResetOnStartup`, but does not trigger any work
- **We implicitly treat this as if reset has executed correctly**, right?

So it's functionally more like "ResetOnStartup = We think the state is 'Reset' after reboot" rather than "ResetOnStartup = We need to trigger a reset operation after reboot"?

Reply:
Yes, that's right, this `RssStatus` is only setting the initial state of the internal state machine, which only allows initialization to proceed if we're currently in a reset state (either `ResetOnStartup` or `Reset` from an actual reset request).

`ResetOnStartup` does sound too much like an action that is going to happen. A couple ideas:

* We could squish these states into one and keep the boolean: `RssStatus::Startup { initialized: bool }`
* Keep the two statuses, but rename `ResetOnStartup` to... something better. Maybe `NotInitializedOnStartup`?

Reply:
Another option for renaming `ResetOnStartup`: `RssStatus::Uninitialized`. That is maybe my favorite? We're not initialized but haven't been reset, either.

Reply:
I also think a state named `Startup` would convey this information - we just booted, we don't know what operations occurred, but we haven't finished initialization.

My only hesitation with the term `Reset` in the name comes from me incorrectly assuming it describes "this state means a reset operation **finished**"

Reply:
Oh, nevermind, I see you used the `InitializedOnStartup` state name, and just calling it `Startup` is a bit ambiguous in comparison

Reply:
Calling it `Uninitialized` seems great.

Reply:
Renamed (and slightly reworked to collapse the two initialized and two uninitialized states down into one) in [cfd2ce7](https://github.com/oxidecomputer/omicron/pull/3351/commits/cfd2ce754f38947c56e49d87a60e4963953c0c42)